diff -up kwin-5.21.0/libkwineffects/KWinEffectsConfig.cmake.in.omv~ kwin-5.21.0/libkwineffects/KWinEffectsConfig.cmake.in
--- kwin-5.21.0/libkwineffects/KWinEffectsConfig.cmake.in.omv~	2021-02-15 19:35:02.317417925 +0100
+++ kwin-5.21.0/libkwineffects/KWinEffectsConfig.cmake.in	2021-02-15 19:35:12.933446301 +0100
@@ -3,7 +3,7 @@
 include(CMakeFindDependencyMacro)
 find_dependency(Qt5Core "@QT_MIN_VERSION@")
 find_dependency(Qt5Gui "@QT_MIN_VERSION@")
-find_dependency(KF5ConfigCore "@KF5_MIN_VERSION@")
+find_dependency(KF5Config "@KF5_MIN_VERSION@")
 find_dependency(KF5CoreAddons "@KF5_MIN_VERSION@")
 find_dependency(KF5WindowSystem "@KF5_MIN_VERSION@")
 
