--- kwin-5.8.4/CMakeLists.txt.omv~	2016-11-23 22:29:59.991580957 +0100
+++ kwin-5.8.4/CMakeLists.txt	2016-11-23 22:31:32.550352012 +0100
@@ -47,6 +47,7 @@ include(ECMInstallIcons)
 include(ECMOptionalAddSubdirectory)
 
 add_definitions(-DQT_DISABLE_DEPRECATED_BEFORE=0 -DQT_USE_QSTRINGBUILDER)
+include_directories("/usr/include/qt5/QtFontDatabaseSupport/5.8.0/")
 
 # require at least gcc 4.8
 if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
--- kwin-5.8.4/plugins/qpa/CMakeLists.txt.omv~	2016-11-23 23:30:43.162494789 +0100
+++ kwin-5.8.4/plugins/qpa/CMakeLists.txt	2016-11-23 23:34:02.349798949 +0100
@@ -25,6 +25,9 @@ target_link_libraries(KWinQpaPlugin
     Qt5PlatformSupport::Qt5PlatformSupport
     ${FONTCONFIG_LIBRARIES}
     ${FREETYPE_LIBRARIES}
+    -lQt5FontDatabaseSupport
+    -lQt5ThemeSupport
+    -lQt5EventDispatcherSupport
 )
 
 if(HAVE_WAYLAND_EGL)
