diff -up kwin-5.21.4/effects/backgroundcontrast/contrast.cpp.4~ kwin-5.21.4/effects/backgroundcontrast/contrast.cpp
--- kwin-5.21.4/effects/backgroundcontrast/contrast.cpp.4~	2021-04-06 14:27:01.000000000 +0200
+++ kwin-5.21.4/effects/backgroundcontrast/contrast.cpp	2021-04-26 03:20:25.176991015 +0200
@@ -256,6 +256,9 @@ bool ContrastEffect::enabledByDefault()
 
     if (gl->isIntel() && gl->chipClass() < SandyBridge)
         return false;
+    // As of Linux 5.12, Mesa 21.1, the effect works, but is very slow (< 5 FPS) on these GPUs
+    if (gl->isLima() || (gl->isPanfrost() && gl->chipClass() <= MaliT8XX) || gl->isVideoCore4() || gl->isVideoCore3D())
+        return false;
     if (gl->isSoftwareEmulation()) {
         return false;
     }
diff -up kwin-5.21.4/effects/blur/blur.cpp.4~ kwin-5.21.4/effects/blur/blur.cpp
--- kwin-5.21.4/effects/blur/blur.cpp.4~	2021-04-06 14:27:01.000000000 +0200
+++ kwin-5.21.4/effects/blur/blur.cpp	2021-04-26 03:17:49.371832827 +0200
@@ -356,6 +356,16 @@ bool BlurEffect::enabledByDefault()
 
     if (gl->isIntel() && gl->chipClass() < SandyBridge)
         return false;
+
+    // As of Linux 5.12, Mesa 21.1, the blur effect works
+    // but is painfully slow (FPS < 5) on Mali and VideoCore
+    if ( gl->isLima() ||
+        (gl->isPanfrost() && gl->chipClass() <= MaliT8XX) ||
+         gl->isVideoCore4() ||
+         gl->isVideoCore3D()) {
+        return false;
+    }
+
     if (gl->isSoftwareEmulation()) {
         return false;
     }
diff -up kwin-5.21.4/libkwineffects/kwinglplatform.cpp.4~ kwin-5.21.4/libkwineffects/kwinglplatform.cpp
--- kwin-5.21.4/libkwineffects/kwinglplatform.cpp.4~	2021-04-06 14:27:01.000000000 +0200
+++ kwin-5.21.4/libkwineffects/kwinglplatform.cpp	2021-04-26 03:19:20.705096243 +0200
@@ -446,23 +446,69 @@ static ChipClass detectQualcommClass(con
     if (ok) {
         if (value >= 100 && value < 200) {
             return Adreno1XX;
-        }
-        if (value >= 200 && value < 300) {
+        } else if (value >= 200 && value < 300) {
             return Adreno2XX;
-        }
-        if (value >= 300 && value < 400) {
+        } else if (value >= 300 && value < 400) {
             return Adreno3XX;
-        }
-        if (value >= 400 && value < 500) {
+        } else if (value >= 400 && value < 500) {
             return Adreno4XX;
-        }
-        if (value >= 500 && value < 600) {
+        } else if (value >= 500 && value < 600) {
             return Adreno5XX;
         }
     }
     return UnknownAdreno;
 }
 
+static ChipClass detectLimaClass(const QByteArray &chipClass)
+{
+
+    if (chipClass.contains("400"))
+        return Mali400;
+    else if (chipClass.contains("450"))
+        return Mali450;
+    else if (chipClass.contains("470"))
+        return Mali470;
+
+    return UnknownLima;
+}
+
+static ChipClass detectPanfrostClass(const QByteArray &chipClass)
+{
+
+    if (chipClass.contains("T720")   ||
+           chipClass.contains("T760"))
+        return MaliT7XX;
+
+    if (chipClass.contains("T820")   ||
+           chipClass.contains("T860"))
+        return MaliT8XX;
+
+    if (chipClass.contains("G31")   ||
+           chipClass.contains("G52") ||
+           chipClass.contains("G72"))
+        return MaliGXX;
+
+    return UnknownPanfrost;
+}
+
+static ChipClass detectVC4Class(const QByteArray &chipClass)
+{
+
+    if (chipClass.contains("2.1"))
+        return VC4_2_1;
+
+    return UnknownVideoCore4;
+}
+
+static ChipClass detectV3DClass(const QByteArray &chipClass)
+{
+
+    if (chipClass.contains("4.2"))
+        return V3D_4_2;
+
+    return UnknownVideoCore3D;
+}
+
 QString GLPlatform::versionToString(qint64 version)
 {
     return QString::fromLatin1(versionToString8(version));
@@ -523,6 +569,14 @@ QByteArray GLPlatform::driverToString8(D
         return QByteArrayLiteral("Qualcomm");
     case Driver_Virgl:
         return QByteArrayLiteral("Virgl (virtio-gpu, Qemu/KVM guest)");
+    case Driver_Lima:
+        return QByteArrayLiteral("Mali (Lima)");
+    case Driver_Panfrost:
+        return QByteArrayLiteral("Mali (Panfrost)");
+    case Driver_VC4:
+        return QByteArrayLiteral("VideoCore IV");
+    case Driver_V3D:
+        return QByteArrayLiteral("VideoCore 3D");
 
     default:
         return QByteArrayLiteral("Unknown");
@@ -604,6 +658,13 @@ QByteArray GLPlatform::chipClassToString
     case Adreno5XX:
         return QByteArrayLiteral("Adreno 5xx series");
 
+    case MaliT7XX:
+        return QByteArrayLiteral("Mali T7xx series");
+    case MaliT8XX:
+        return QByteArrayLiteral("Mali T8xx series");
+    case MaliGXX:
+        return QByteArrayLiteral("Mali Gxx series");
+
     default:
         return QByteArrayLiteral("Unknown");
     }
@@ -787,6 +848,26 @@ void GLPlatform::detect(OpenGLPlatformIn
         m_chipClass = detectQualcommClass(m_renderer);
     }
 
+    else if (m_renderer.contains("Mali")) {
+        m_driver = Driver_Lima;
+        m_chipClass = detectLimaClass(m_renderer);
+    }
+
+    else if (m_renderer.contains("Panfrost")) {
+        m_driver = Driver_Panfrost;
+        m_chipClass = detectPanfrostClass(m_renderer);
+    }
+
+    else if (m_renderer.startsWith("VC4 ")) {
+        m_driver = Driver_VC4;
+	m_chipClass = detectVC4Class(m_renderer);
+    }
+
+    else if(m_renderer.startsWith("V3D ")) {
+        m_driver = Driver_V3D;
+	m_chipClass = detectV3DClass(m_renderer);
+    }
+
     else if (m_renderer == "Software Rasterizer") {
         m_driver = Driver_Swrast;
     }
@@ -912,6 +993,7 @@ void GLPlatform::detect(OpenGLPlatformIn
         else if (m_renderer == "virgl") {
             m_driver = Driver_Virgl;
         }
+
     }
 
     // Driver/GPU specific features
@@ -984,6 +1066,31 @@ void GLPlatform::detect(OpenGLPlatformIn
         }
     }
 
+    if (isLima()) {
+        m_recommendedCompositor = OpenGL2Compositing;
+        // As of Linux 5.12, Mesa 21.1, GLSL works but causes dramatic FPS drop on this GPU
+        m_supportsGLSL = false;
+    }
+
+    if (isPanfrost()) {
+        m_recommendedCompositor = OpenGL2Compositing;
+
+        if (m_chipClass <= MaliT8XX) {
+            // As of Linux 5.12, Mesa 21.1, GLSL works but causes dramatic FPS drop on this GPU
+            m_supportsGLSL = false;
+        }
+    }
+
+    if (isVideoCore4()) {
+        // As of Linux 5.12, Mesa 21.1, OpenGL works, but is much slower than XRender
+        m_recommendedCompositor = XRenderCompositing;
+    }
+
+    if (isVideoCore3D()) {
+        // As of Linux 5.12, Mesa 21.1, OpenGL works, but is much slower than XRender
+        m_recommendedCompositor = XRenderCompositing;
+    }
+
     if (isMesaDriver() && platformInterface == EglPlatformInterface) {
         // According to the reference implementation in
         // mesa/demos/src/egl/opengles1/texture_from_pixmap
@@ -1204,6 +1311,26 @@ bool GLPlatform::isAdreno() const
     return m_chipClass >= Adreno1XX && m_chipClass <= UnknownAdreno;
 }
 
+bool GLPlatform::isLima() const
+{
+    return m_chipClass >= Mali400 && m_chipClass <= UnknownLima;
+}
+
+bool GLPlatform::isPanfrost() const
+{
+    return m_chipClass >= MaliT7XX && m_chipClass <= UnknownPanfrost;
+}
+
+bool GLPlatform::isVideoCore4() const
+{
+    return m_chipClass >= VC4_2_1 && m_chipClass <= UnknownVideoCore4;
+}
+
+bool GLPlatform::isVideoCore3D() const
+{
+    return m_chipClass >= V3D_4_2 && m_chipClass <= UnknownVideoCore3D;
+}
+
 const QByteArray &GLPlatform::glRendererString() const
 {
     return m_renderer;
diff -up kwin-5.21.4/libkwineffects/kwinglplatform.h.4~ kwin-5.21.4/libkwineffects/kwinglplatform.h
--- kwin-5.21.4/libkwineffects/kwinglplatform.h.4~	2021-04-06 14:27:01.000000000 +0200
+++ kwin-5.21.4/libkwineffects/kwinglplatform.h	2021-04-26 03:15:40.025053611 +0200
@@ -93,6 +93,10 @@ enum Driver {
     Driver_Qualcomm,
     Driver_RadeonSI,
     Driver_Virgl,
+    Driver_Lima,
+    Driver_Panfrost,
+    Driver_VC4,
+    Driver_V3D,
     Driver_Unknown
 };
 
@@ -142,6 +146,28 @@ enum ChipClass {
     Adreno5XX,              // GLES3.1 CL2.0 DX11.2
     UnknownAdreno = 3999,
 
+    // Lima Mali
+    // from https://docs.mesa3d.org/drivers/lima.html
+    Mali400       = 4000,
+    Mali450,
+    Mali470,
+    UnknownLima = 4999,
+
+    // Panfrost Mali
+    // from https://docs.mesa3d.org/drivers/panfrost.html
+    MaliT7XX      = 5000,   // GLES2.0
+    MaliT8XX,               // GLES3.0
+    MaliGXX,                // GLES3.0
+    UnknownPanfrost = 5999,
+
+    // Broadcom VideoCore IV (e.g. Raspberry Pi 0 to 3), GLES 2.0/2.1 with caveats
+    VC4_2_1       = 6000, // Found in Raspberry Pi 3B+
+    UnknownVideoCore4 = 6999,
+
+    // Broadcom VideoCore 3D (e.g. Raspberry Pi 4, Raspberry Pi 400)
+    V3D_4_2       = 7000, // Found in Raspberry Pi 400
+    UnknownVideoCore3D = 7999,
+
     UnknownChipClass = 99999
 };
 
@@ -287,6 +313,30 @@ public:
     bool isVirgl() const;
 
     /**
+     * @returns @c true if the GPU is a Mali GPU supported by the Lima driver (Mali 400, 450)
+     * @since 5.22
+     **/
+    bool isLima() const;
+
+    /**
+     * @returns @c true if the GPU is a Mali GPU supported by the Panfrost driver (Mali T7xx, T8xx, Gxx)
+     * @since 5.22
+     **/
+    bool isPanfrost() const;
+
+    /**
+     * @returns @c true if the GPU is a Broadcom VideoCore IV (e.g. Raspberry Pi 0 to 3)
+     * @since 5.22
+     **/
+    bool isVideoCore4() const;
+
+    /**
+     * @returns @c true if the GPU is a Broadcom VideoCore 3D (e.g. Raspberry Pi 4, 400)
+     * @since 5.22
+     **/
+    bool isVideoCore3D() const;
+
+    /**
      * @returns the GL_VERSION string as provided by the driver.
      * @since 4.9
      */
