diff -up kwin-5.17.0/plugins/qpa/eglhelpers.cpp.omv~ kwin-5.17.0/plugins/qpa/eglhelpers.cpp
--- kwin-5.17.0/plugins/qpa/eglhelpers.cpp.omv~	2019-10-10 21:39:15.923492930 +0200
+++ kwin-5.17.0/plugins/qpa/eglhelpers.cpp	2019-10-10 21:48:43.385553399 +0200
@@ -19,10 +19,10 @@ You should have received a copy of the G
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *********************************************************************/
 
-#include "eglhelpers.h"
-
 #include <logging.h>
 
+#include "eglhelpers.h"
+
 #include <QOpenGLContext>
 
 namespace KWin
diff -up kwin-5.17.0/plugins/qpa/eglhelpers.h.omv~ kwin-5.17.0/plugins/qpa/eglhelpers.h
--- kwin-5.17.0/plugins/qpa/eglhelpers.h.omv~	2019-10-10 21:47:10.821104187 +0200
+++ kwin-5.17.0/plugins/qpa/eglhelpers.h	2019-10-10 21:48:21.763449653 +0200
@@ -24,6 +24,11 @@ along with this program.  If not, see <h
 #include <epoxy/egl.h>
 #include "fixqopengl.h"
 
+// Fun with defines from X11 headers #define-ing stuff
+// that Qt uses as enum members...
+#undef None
+#undef CursorShape
+
 #include <QSurfaceFormat>
 
 namespace KWin
diff -up kwin-5.17.0/plugins/qpa/offscreensurface.cpp.omv~ kwin-5.17.0/plugins/qpa/offscreensurface.cpp
--- kwin-5.17.0/plugins/qpa/offscreensurface.cpp.omv~	2019-10-10 21:51:20.418288355 +0200
+++ kwin-5.17.0/plugins/qpa/offscreensurface.cpp	2019-10-10 21:53:52.682975860 +0200
@@ -18,13 +18,14 @@ You should have received a copy of the G
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *********************************************************************/
 
+#include <QOffscreenSurface>
+#include <QDataStream>
+
 #include "offscreensurface.h"
 #include "eglhelpers.h"
 #include "main.h"
 #include "platform.h"
 
-#include <QOffscreenSurface>
-
 namespace KWin
 {
 namespace QPA
diff -up kwin-5.17.0/plugins/qpa/offscreensurface.h.omv~ kwin-5.17.0/plugins/qpa/offscreensurface.h
--- kwin-5.17.0/plugins/qpa/offscreensurface.h.omv~	2019-10-10 21:49:44.061841036 +0200
+++ kwin-5.17.0/plugins/qpa/offscreensurface.h	2019-10-10 21:54:24.870118662 +0200
@@ -24,6 +24,17 @@ along with this program.  If not, see <h
 #include <epoxy/egl.h>
 #include "fixqopengl.h"
 
+// fun with X11 headers #define-ing stuff Qt uses as enum members
+#undef FontChange
+#undef CursorShape
+#undef Expose
+#undef KeyPress
+#undef KeyRelease
+#undef FocusIn
+#undef FocusOut
+#undef Bool
+#undef Always
+
 #include <qpa/qplatformoffscreensurface.h>
 
 namespace KWin
